### ISSUE TITLE
identity_op seems confused sometimes

### ISSUE THREAD

**Issue by fabricedesre** [Manishearth/rust-clippy#969]
Mostly failing on #[derive(...)] expressions (code is at https://github.com/fxbox/foxbox/blob/1095a39b22db140df31c206a3ac5ded12f13ddf1/src/adapters/webpush/mod.rs):

```rust
src/adapters/webpush/mod.rs:47:1: 47:52 warning: the operation is ineffective. Consider reducing it to `#[derive(Debug, PartialEq, Serialize, Deserialize)]`
src/adapters/webpush/mod.rs:47 #[derive(Debug, PartialEq, Serialize, Deserialize)]
                               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
src/main.rs:16:9: 16:20 note: lint level defined here
src/main.rs:16 #![warn(identity_op)]
                       ^~~~~~~~~~~
src/adapters/webpush/mod.rs:47:1: 47:52 help: for further information visit https://github.com/Manishearth/rust-clippy/wiki#identity_op
src/adapters/webpush/mod.rs:54:1: 54:52 warning: the operation is ineffective. Consider reducing it to `#[derive(Debug, PartialEq, Serialize, Deserialize)]`
src/adapters/webpush/mod.rs:54 #[derive(Debug, PartialEq, Serialize, Deserialize)]
                               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
src/main.rs:16:9: 16:20 note: lint level defined here
src/main.rs:16 #![warn(identity_op)]
                       ^~~~~~~~~~~
src/adapters/webpush/mod.rs:54:1: 54:52 help: for further information visit https://github.com/Manishearth/rust-clippy/wiki#identity_op
src/adapters/webpush/mod.rs:67:1: 67:52 warning: the operation is ineffective. Consider reducing it to `#[derive(Debug, PartialEq, Serialize, Deserialize)]`
src/adapters/webpush/mod.rs:67 #[derive(Debug, PartialEq, Serialize, Deserialize)]
                               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
src/main.rs:16:9: 16:20 note: lint level defined here
src/main.rs:16 #![warn(identity_op)]
```

**Comment by Mrmaxmeier** [comment]
I've encountered a similar warning recently:
```
warning: the operation is ineffective. Consider reducing it to `#[derive(Serialize)]`, #[warn(identity_op)] on by default
 --> src/main.rs:5:1
  |>
5 |> #[derive(Serialize)]
  |> ^^^^^^^^^^^^^^^^^^^^
help: for further information visit https://github.com/Manishearth/rust-clippy/wiki#identity_op
```

Here's a minimal snippet to reproduce the warning:
```rust
#![feature(custom_derive, plugin)]
#![plugin(serde_macros)]
extern crate serde_json;

#[derive(Serialize)]
struct A { b: u8 }

fn main() {}
```

Similar to how [macros get skipped](https://github.com/Manishearth/rust-clippy/blob/master/clippy_lints/src/identity_op.rs#L31) for the `identity_op` lint, I guess derived code should be skipped too.

**Comment by Manishearth** [comment]
in_macro should be skipping `derive`, too. Not sure why it isn't working. @oli-obk, any idea? The handling of expansion spans recently changed.

**Comment by oli-obk** [comment]
I think this is because serde changes all spans to the span of the `Serialize` (or the `#[derive(Serialize)]` depending on your rustc version). This span is "obviously" not inside a macro...

**Comment by Manishearth** [comment]
Oh, I see. That's a serde bug I guess?

**Comment by oli-obk** [comment]
yea

**Comment by dtolnay** [comment]
This lint is catching the way we count up the number of serialized fields in a struct, accounting for ones that are skipped at runtime: [serde_codegen/src/ser.rs](https://github.com/serde-rs/serde/blob/8a09f05644ad286e5a18c714ec1af58ed90d1d8b/serde_codegen/src/ser.rs#L810). In the short term I can change that code to omit the `0+`.

@oli-obk filed https://github.com/serde-rs/serde/issues/350 *"serde_macros spans need add expansion info."* Can anyone point me to docs or examples of how this is supposed to work? I am not sure where to begin on this one.

**Comment by dtolnay** [comment]
We already surround everything we generate with `#[automatically_derived]` [here](https://github.com/serde-rs/serde/blob/5cd8212a61e57c56ec0503e9640bbcb485ba855c/serde_codegen/src/ser.rs#L86). Instead of changing anything in Serde, how about whitelisting some of the lints inside of automatically_derived? I imagine many lints we want to keep inside of generated code (like inline_always) but the ones that are focused on human-readability of code (identity_op) should be whitelisted because we want the generat***ing*** code to be human-readable, not the generat***ed*** code.

**Comment by dtolnay** [comment]
I merged a fix and it will be included in the next serde_macros release.

**Comment by dtolnay** [comment]
I released serde_macros 0.7.13 which fixes this issue.

**Comment by Manishearth** [comment]
Thanks!
